### PR TITLE
Possibility to lowercase/uppercase matched user name

### PR DESCRIPTION
**Pull request by iirekm**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 1]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
...
                 @JsonProperty("user") Optional<String> user,
-                @JsonProperty("allow") Optional<Boolean> allow)
+                @JsonProperty("allow") Optional<Boolean> allow,
+                @JsonProperty("case") Optional<Case> kase)
         {
             this(pattern,
                     requireNonNull(user, "user is null").orElse("$1"),
-                    requireNonNull(allow, "allow is null").orElse(TRUE));
+                    requireNonNull(allow, "allow is null").orElse(TRUE),
+                    requireNonNull(kase, "kase is null").orElse(Case.KEEP));
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:131]: Case? Like class.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 2]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
...
         }
 
-        public Rule(String pattern, String user, boolean allow)
+        public Rule(String pattern, String user, boolean allow, Case kase)
         {
             this.pattern = Pattern.compile(requireNonNull(pattern, "pattern is null"));
             this.user = requireNonNull(user, "user is null");
             this.allow = allow;
+            this.kase = kase;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:139]: Use require non null

File: presto-main/src/test/java/io/prestosql/server/security/TestUserMapping.java [hunk 3]
```diff
@@ -70,29 +70,29 @@ public void testSimplePatternRule()
...
+        UserMapping emptyMapping = new UserMapping(ImmutableList.of(new Rule("(.*?)@.*", "", false, UserMapping.Case.KEEP)));
         assertThrows(UserMappingException.class, () -> emptyMapping.mapUser("test@example.com"));
     }
 
     @Test
     public void testMultipleRule()
             throws Exception
     {
-        UserMapping userMapping = new UserMapping(ImmutableList.of(new Rule("test@example.com", "", false), new Rule("(.*?)@example.com")));
+        UserMapping userMapping = new UserMapping(ImmutableList.of(new Rule("test@example.com", "", false, UserMapping.Case.KEEP), new Rule("(.*?)@example.com")));
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/security/TestUserMapping.java:95]: Static import

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 4]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
         }
     }
 
+    enum Case
+    {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:90]: We also need to update documentation

Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:90]: where is it?

File: presto-main/src/test/java/io/prestosql/server/security/TestUserMapping.java (hunk 3 above)
Comment (iirekm) [reply presto-main/src/test/java/io/prestosql/server/security/TestUserMapping.java:95]: can't somebody configure checkstyle / sonar / intellij rules for that, once and for all? This is what intellij does by default after typing KEEP

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 1 above)
Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:131]: parameter name is `kase` because `case` is a keyword. The thing commonly used in other projects is something like `_case` or `case_` but here checkstyle disallows it, so changing spelling is second commonly used thing (like `class` -> `klass` or `clazz`)

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 2 above)
Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:139]: I forgot

But anyway, isn't this code too defensive? Using `requireNotNull` everywhere adds a lot of noise to code. Only 4 lines (field declarations) in this class are meaningful, the rest is ugly boilerplate code!

It's enough to use `@NotNull` annotation, IntelliJ recognizes it, and can be configured to produce compile error when you try to to pass a nullable value to `@NotNull` parameter.

Another alternative: use immutables.org annotation processor. It can generate all this mess automatically, eg recognizes `@NotNull` annotation and generates appropriate `requireNonNull` from it.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:90]: https://github.com/prestosql/presto/blob/master/presto-docs/src/main/sphinx/security/user-mapping.rst

File: presto-main/src/test/java/io/prestosql/server/security/TestUserMapping.java (hunk 3 above)
Comment (kokosing) [reply presto-main/src/test/java/io/prestosql/server/security/TestUserMapping.java:95]: It is hard to have a rule for a class that you have just added. Also I would prefer to not have such rule for `Case` class as in this context it is clear what it is, but in other it might not be.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:139]: > `@NotNul`

This is also a boilreplate. In general most variables are non null. So I would prefer to configure compiler to fail if `@Nullable` is missing. Is it possible.

Also this is a discussion about project convention which way out of the scope of this pull request.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:131]: uh... sorry auto correction fix `caze` to `case`. I wanted to say:

`caze`? like `clazz`? `kase` looks like a typo to me. Also I am under impression that it is more common in languages that `z` sounds closer to `s` than `k` to `c`.

I don't have strong preference. Up to you.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 5]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
...
+            return transformation.apply(value);
+        }
+    }
+
     public static final class Rule
     {
         private final Pattern pattern;
         private final String user;
         private final boolean allow;
+        private final Case kase;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:114]: Alternatively, `capitalization`

Comment (electrum) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:114]: Or `userCase`, since this applies to the username.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:131]: I am not sure about this regexp. Please open such pull request and we could continue the discussion there. How about applying: https://github.com/prestosql/presto/pull/4736/files#r467496420?

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/util/JsonUtils.java [hunk 6]
```diff
@@ -42,7 +43,8 @@ private JsonUtils() {}
         try {
             byte[] json = Files.readAllBytes(path);
             ObjectMapper mapper = new ObjectMapperProvider().get()
-                    .enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+                    .enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+                    .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS);
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/util/JsonUtils.java:47]: Please extract this to separate commit and add a test for that.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 7]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
...
+            this(pattern, "$1", true, Case.KEEP);
         }
 
         @JsonCreator
         public Rule(
                 @JsonProperty("pattern") String pattern,
                 @JsonProperty("user") Optional<String> user,
-                @JsonProperty("allow") Optional<Boolean> allow)
+                @JsonProperty("allow") Optional<Boolean> allow,
+                @JsonProperty("case") Optional<Case> kase)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:126]: Let's call this `userCase` everywhere instead.

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 8]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
         }
     }
 
+    enum Case
+    {
+        KEEP(Function.identity()),
+        LOWER(value -> value.toLowerCase(Locale.ENGLISH)),
+        UPPER(value -> value.toUpperCase(Locale.ENGLISH)),
+        /**/;
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:94]: No need for this here, unless we think we will be adding more later. The purpose of this construct is to have cleaner diffs when adding future items (as the previously last item won't be disturbed).

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 9]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
...
+
+        private final Function<String, String> transformation;
+
+        Case(Function<String, String> transformation)
+        {
+            this.transformation = requireNonNull(transformation, "transformation is null");
+        }
+
+        public final String transform(String value)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:103]: No need for final, since enums are sealed

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 10]
```diff
@@ -84,33 +86,57 @@ public UserMappingRules(
...
+    enum Case
+    {
+        KEEP(Function.identity()),
+        LOWER(value -> value.toLowerCase(Locale.ENGLISH)),
+        UPPER(value -> value.toUpperCase(Locale.ENGLISH)),
+        /**/;
+
+        private final Function<String, String> transformation;
+
+        Case(Function<String, String> transformation)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:98]: Consider the simpler (albeit slightly longer) pattern of implementing a method (the IDE folds it nicely)
```java
enum Case
{
    KEEP {
        @Override
        public String transform(String value)
        {
            return value;
        }
    },
    LOWER {
        @Override
        public String transform(String value)
        {
            return value.toLowerCase(Locale.ENGLISH);
        }
    },
    UPPER {
        @Override
        public String transform(String value)
        {
            return value.toLowerCase(Locale.ENGLISH);
        }
    };

    public abstract String transform(String value);
}

File: presto-docs/src/main/sphinx/security/user-mapping.rst [hunk 11]
```diff
@@ -41,6 +41,13 @@ following fields:
 * ``user`` (optional): replacement string to substitute against pattern.
   The default value is ``$1``.
 * ``allow`` (optional): boolean indicating if the authentication should be allowed.
+* ``case`` (optional): one of:
```
Comment (electrum) [inline presto-docs/src/main/sphinx/security/user-mapping.rst:44]: Thanks for documenting this. Some comments:

* List items need to be indented two spaces. The extra indentation causes it to be treated as a quoted block.
* Use `*` for list items to be consistent with the other list items (above).
* The "default value" text below is hard to see, since it's not aligned with the list. We can make it part of the first list item.
* e.g. means "for example" and thus should have a comma afterwards
* The lower/upper items are long enough to wrap in the default HTML rendering, which makes them harder to read. We can shorten them a bit so that they don't wrap.

Here's the result of the above changes:

```
* ``case`` (optional): one of:

  * ``keep`` - keep matched user name as is (default behavior)
  * ``lower`` - lowercase matched user name, e.g., ``Admin`` or ``ADMIN`` will become ``admin``
  * ``upper`` - uppercase matched user name, e.g., ``admin`` or ``Admin`` will become ``ADMIN``
```

Rendered HTML:

Screen Shot 2020-08-14 at 9 46 21 AM

Comment (iirekm) [reply presto-docs/src/main/sphinx/security/user-mapping.rst:44]: done

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 10 above)
Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java]: done

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 5 above)
Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java]: done

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java (hunk 7 above)
Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/server/security/UserMapping.java]: done

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/util/JsonUtils.java (hunk 6 above)
Comment (iirekm) [reply presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/util/JsonUtils.java]: done

File: presto-main/src/main/java/io/prestosql/server/security/UserMapping.java [hunk 12]
```diff
@@ -84,33 +86,64 @@ public UserMappingRules(
...
+            }
+        },
+        LOWER {
+            @Override
+            public String transform(String value)
+            {
+                return value.toLowerCase(ENGLISH);
+            }
+        },
+        UPPER {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/server/security/UserMapping.java:105]: nice syntax, I haven't used enums that way.

File: presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/util/JsonUtils.java [hunk 13]
```diff
@@ -41,12 +43,20 @@ private JsonUtils() {}
...
-            return mapper.readValue(json, javaType);
+            return parseJson(json, javaType);
         }
         catch (IOException e) {
             throw new IllegalArgumentException(format("Invalid JSON file '%s' for '%s'", path, javaType), e);
         }
     }
+
+    @VisibleForTesting
+    static <T> T parseJson(byte[] jsonBytes, Class<T> javaType)
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/util/JsonUtils.java:54]: Please extract this to separate commit.

File: presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/util/TestJsonUtils.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        @JsonProperty
+        private TestEnum testEnum;
+    }
+
+    @Test
+    public void testLowercaseEnum()
+            throws IOException
+    {
+        TestObject parsed = parseJson("{\"testEnum\": \"option_a\"}".getBytes(US_ASCII), TestObject.class);
```
Comment (kokosing) [inline presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/util/TestJsonUtils.java:44]: please also test mixed and upper case.

Comment (iirekm) [reply presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/util/TestJsonUtils.java:44]: you told that we should support only lower case

uppercase/mixed case would also work (there's no switch in Jackson to enable only lowercase, there's only `ACCEPT_CASE_INSENSITIVE_ENUMS`), but I would say it should stay undocumented, untested and unsupported, if somebody uses it and some day it stops working, it's their fault

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
Please ping me when automation completes, so I can merge it. Thanks!

**Comment by iirekm** [comment]
build at least passed after rebase - ready to merge

**Comment by kokosing** [comment]
Merged, thanks!
